integration-tests/tests/src/tests/sync/sync-session.ts

```diff
@@ -0,0 +1,1415 @@
+      const { user, config } = await getSyncConfWithUser(this.app, partition);
+      config.schemaVersion = 1;
+
+      // Open the Realm with a schema version of 1, then immediately close it.
+      // This verifies that Realm.open doesn't hit issues when the schema version
+      // of an existing, local Realm is different than the one passed in the configuration.
+      const realm = new Realm(config);
+      realm.close();
+
+      config.schemaVersion = 2;
+      const realm2 = await Realm.open(config);
+      const actualObjectsCount = realm2.objects("Dog").length;
+      expect(actualObjectsCount).equals(expectedObjectsCount);
+      const session = realm2.syncSession;
+      expect(session).instanceOf(Realm.App.Sync.Session);
+      expect(session?.user.id).equals(user.id);
+      //@ts-expect-error comparing undefined with undefined, url does not exist on config.
+      expect(session?.config.url).equals(config.sync?.url);
+      expect(session?.config.partitionValue).equals(config.sync?.partitionValue);
+      expect(session?.config.user.id).equals(config.sync?.user.id);
+      expect(session?.state).equals("active");
+    });
+  });
+
+  describe("error handling", () => {
+    afterEach(() => Realm.clearTestState());
+    it("can simulate an error", async function (this: AppContext) {
+      const partition = generatePartition();
+      const credentials = Realm.Credentials.anonymous();
+      return this.app.logIn(credentials).then((user) => {
+        return new Promise((resolve, _reject) => {
+          const config = getSyncConfiguration(user, partition);
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          config.sync!.onError = (_, error: any) => {
+            try {
+              expect(error.message).equals("simulated error");
+              expect(error.code).equals(123);
+              resolve();
+            } catch (e) {
+              _reject(e);
+            }
+          };
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          config.sync!.clientReset = {
+            //@ts-expect-error TYPEBUG: can't apply enums.
+            mode: "manual",
+          };
+          const realm = new Realm(config);
+          const session = realm.syncSession;
+          //@ts-expect-error using internal method.
+          session._simulateError(123, "simulated error", "realm::sync::ProtocolError", false);
+        });
+      });
+    });
+  });
+
+  describe("progress notification", () => {
+    afterEach(() => Realm.clearTestState());
+    it("is called", async function (this: AppContext) {
+      const partition = generatePartition();
+      const { config } = await getSyncConfWithUser(this.app, partition);
+      let progressCalled = false;
+      await Promise.race([
+        Realm.open(config).progress(() => {
+          progressCalled = true;
+        }),
+        new Promise(() => {
+          sleep(5000, false);
+        }),
```

Comment:
No need to wrap this in a promise constructor.
Also, since the callback body is neither awaiting nor returning the promise, this is effectively a promise that never resolves but calls `sleep` which might become an unhandled promise rejection.

```suggestion
        sleep(5000, false),
```